lib/cn/csched_sp3_work.c

```diff
@@ -1049,10 +1049,16 @@ sp3_work(
     if (w->cw_action == CN_ACTION_SPILL) {
         w->cw_sgen = ++tree->ct_sgen;
     } else if (w->cw_action == CN_ACTION_JOIN) {
+        uint64_t workid = 0;
 
-        /* TODO: Should we set the work ID of each kvset in cw_join?
-         */
         w->cw_join = list_prev_entry(tn, tn_link);
+
+        list_for_each_entry_reverse(le, &w->cw_join->tn_kvset_list, le_link) {
+            if (workid == 0)
+                workid = kvset_get_dgen(le->le_kvset);
+
+            kvset_set_workid(le->le_kvset, workid);
+        }
```

Comment:
Maybe:
uint64_t workid = kvset_get_dgen(*mark);
then lose the if (worked == 0) stuff...

Reply:
I thought we set 'mark' on w->cw_node for node join? In this code segment, we set the workid for kvsets in the left/source node, i.e., w->cw_join.

This will change for zspill as we need to mark the kvsets in the source node and only those marked will be moved to the target node.

Reply:
my bad